### ISSUE TITLE
opt: gracefully handle evaluation errors in partial index predicates

### ISSUE THREAD

**Issue by mgartner** [cockroachdb/cockroach#50214]
Errors when evaluating partial index predicates result in panics.

```
> set experimental_partial_indexes=on;

> create table t (a int, b int, index (a) where a / b > 10);

> insert into t values (1, 0);
*
* ERROR: [n1,client=127.0.0.1:51504,hostssl,user=root] a SQL panic has occurred while executing the following statement:
* INSERT INTO t VALUES (1, 0)
*
*
* ERROR: [n1,client=127.0.0.1:51504,hostssl,user=root] a panic has occurred!
*
panic: runtime error: index out of range [3] with length 3 [recovered]
        panic: runtime error: index out of range [3] with length 3

...
```

```
Build Tag:    v20.2.0-alpha.00000000-2311-g4702d62092
Build Time:   2020/06/15 17:01:50
Distribution: CCL
Platform:     darwin amd64 (x86_64-apple-darwin19.4.0)
Go Version:   go1.13.4
C Compiler:   4.2.1 Compatible Apple LLVM 11.0.0 (clang-1100.0.33.17)
Build SHA-1:  4702d620923e9b5d13a8f4c2a833dcaf5bd40e23
Build Type:   development
```

**Comment by mgartner** [comment]
I think this may be fixed when #51690 is fixed.

**Comment by mgartner** [comment]
Fixed by #51789.
